File: app/src/main/java/com/dhpcs/liquidity/activity/GamesActivity.java
package com.dhpcs.liquidity.activity;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import com.dhpcs.liquidity.R;
import com.dhpcs.liquidity.fragment.AddGameBottomSheetDialogFragment;
import com.dhpcs.liquidity.fragment.CreateGameDialogFragment;
import com.dhpcs.liquidity.fragment.GamesFragment;
import com.dhpcs.liquidity.model.ZoneId;
import com.google.zxing.integration.android.IntentIntegrator;
import com.google.zxing.integration.android.IntentResult;

import java.util.Collections;
import java.util.Currency;

public class GamesActivity extends AppCompatActivity
        implements AddGameBottomSheetDialogFragment.Listener,
        CreateGameDialogFragment.Listener,
        GamesFragment.Listener {

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        IntentResult result = IntentIntegrator.parseActivityResult(requestCode, resultCode, data);
        if (result != null) {
            String contents = result.getContents();
            if (contents != null) {
                startActivity(
                        new Intent(
                                GamesActivity.this,
                                BoardGameActivity.class
                        ).putExtra(
                                BoardGameActivity.EXTRA_ZONE_ID,
                                new ZoneId(contents)
                        )
                );
            }
        } else {
            super.onActivityResult(requestCode, resultCode, data);
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_games);

        Toolbar toolbar = findViewById(R.id.toolbar);

        setSupportActionBar(toolbar);

        View floatingActionButtonAddGame = findViewById(R.id.floatingactionbutton_add_game);
        assert floatingActionButtonAddGame != null;
        floatingActionButtonAddGame.setOnClickListener(v -> AddGameBottomSheetDialogFragment.newInstance()
                .show(
                        getSupportFragmentManager(),
                        AddGameBottomSheetDialogFragment.TAG
                ));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.games_toolbar, menu);
        return true;
    }

    @Override
    public void onGameClicked(long gameId, ZoneId zoneId, String gameName) {
        startActivity(
                new Intent(
                        GamesActivity.this,
                        BoardGameActivity.class
                ).putExtra(
                        BoardGameActivity.EXTRA_GAME_ID,
                        gameId
                ).putExtra(
                        BoardGameActivity.EXTRA_ZONE_ID,
                        zoneId
                ).putExtra(
                        BoardGameActivity.EXTRA_GAME_NAME,
                        gameName
                )
        );
    }

    @Override
    public void onGameDetailsEntered(String name, Currency currency) {
        startActivity(
                new Intent(
                        GamesActivity.this,
                        BoardGameActivity.class
                ).putExtra(
                        BoardGameActivity.EXTRA_CURRENCY,
                        currency
                ).putExtra(
                        BoardGameActivity.EXTRA_GAME_NAME,
                        name
                )
        );
    }

    @Override
    public void onJoinGameClicked() {
        new IntentIntegrator(GamesActivity.this)
                .setCaptureActivity(JoinGameActivity.class)
                .setDesiredBarcodeFormats(
                        Collections.singleton("QR_CODE")
                )
                .setBeepEnabled(false)
                .setOrientationLocked(false)
                .initiateScan();
    }

    @Override
    public void onNewGameClicked() {
        CreateGameDialogFragment.newInstance()
                .show(
                        getSupportFragmentManager(),
                        CreateGameDialogFragment.TAG
                );
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_preferences:
                startActivity(
                        new Intent(
                                GamesActivity.this,
                                PreferencesActivity.class
                        )
                );
                return true;
            case R.id.action_about:
                startActivity(
                        new Intent(
                                GamesActivity.this,
                                AboutActivity.class
                        )
                );
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

}

File: app/src/main/java/com/dhpcs/liquidity/activity/GamesActivity.kt
package com.dhpcs.liquidity.activity

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.dhpcs.liquidity.R
import com.dhpcs.liquidity.fragment.AddGameBottomSheetDialogFragment
import com.dhpcs.liquidity.fragment.CreateGameDialogFragment
import com.dhpcs.liquidity.fragment.GamesFragment
import com.google.zxing.integration.android.IntentIntegrator
import java.util.*

class GamesActivity :
        AppCompatActivity(),
        AddGameBottomSheetDialogFragment.Companion.Listener,
        CreateGameDialogFragment.Companion.Listener,
        GamesFragment.Companion.Listener {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_games)

        val toolbar = findViewById<Toolbar>(R.id.toolbar)

        setSupportActionBar(toolbar)

        val floatingActionButtonAddGame = findViewById<View>(R.id.floatingactionbutton_add_game)!!
        floatingActionButtonAddGame.setOnClickListener {
            AddGameBottomSheetDialogFragment.newInstance().show(
                    supportFragmentManager,
                    AddGameBottomSheetDialogFragment.TAG
            )
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.games_toolbar, menu)
        return true
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        val result = IntentIntegrator.parseActivityResult(requestCode, resultCode, data)
        when {
            result != null -> {
                val contents = result.contents
                if (contents != null) {
                    startActivity(
                            Intent(this, BoardGameActivity::class.java)
                                    .putExtra(BoardGameActivity.EXTRA_ZONE_ID, contents)
                    )
                }
            }
            else ->
                super.onActivityResult(requestCode, resultCode, data)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_preferences -> {
                startActivity(Intent(this, PreferencesActivity::class.java))
                return true
            }
            R.id.action_about -> {
                startActivity(Intent(this, AboutActivity::class.java))
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onNewGameClicked() {
        CreateGameDialogFragment.newInstance().show(
                supportFragmentManager,
                CreateGameDialogFragment.TAG
        )
    }

    override fun onGameDetailsEntered(name: String, currency: Currency) {
        startActivity(
                Intent(this, BoardGameActivity::class.java)
                        .putExtra(BoardGameActivity.EXTRA_CURRENCY, currency)
                        .putExtra(BoardGameActivity.EXTRA_GAME_NAME, name)
        )
    }

    override fun onJoinGameClicked() {
        IntentIntegrator(this)
                .setCaptureActivity(JoinGameActivity::class.java)
                .setDesiredBarcodeFormats(setOf("QR_CODE"))
                .setBeepEnabled(false)
                .setOrientationLocked(false)
                .initiateScan()
    }

    override fun onGameClicked(gameId: Long, zoneId: String, gameName: String?) {
        startActivity(
                Intent(this, BoardGameActivity::class.java)
                        .putExtra(BoardGameActivity.EXTRA_GAME_ID, gameId)
                        .putExtra(BoardGameActivity.EXTRA_ZONE_ID, zoneId)
                        .putExtra(BoardGameActivity.EXTRA_GAME_NAME, gameName)
        )
    }

}
